File: java/src/raft/ClienteCluster.java
package raft;

import java.io.*;
import java.net.*;
import java.util.*;

// Cliente para hablar con el cluster Raft desde cualquier lenguaje/proceso.
// Lo usa el Servidor de Testeo (para insertar detecciones) y el Cliente Vigilante
// (para leer el registro). Maneja el REDIRECT: si contacto un seguidor, me manda
// al lider y reintento alli.
public class ClienteCluster {

    private final List<Par> nodos;
    private Par ultimoLider;   // recuerdo al lider para no reintentar desde cero

    public ClienteCluster(List<Par> nodos) {
        this.nodos = nodos;
    }

    // Inserta una deteccion. Devuelve true si el cluster la acepto.
    public boolean insertar(String tipo, String imagenRef, String fechaHora, String camara) {
        String cmd = tipo + "," + imagenRef + "," + fechaHora + "," + camara;
        String resp = pedir(Protocolo.unir(Protocolo.NUEVA_DETECCION, cmd));
        return resp != null && resp.equals(Protocolo.OK);
    }

    // Lee el registro cometido. Devuelve la lista de detecciones (cada una es el
    // string  tipo,imagenRef,fechaHora,camara ).
    public List<String> leerRegistro() {
        String resp = pedir(Protocolo.LEER_REGISTRO);
        List<String> out = new ArrayList<>();
        if (resp == null) return out;
        String[] c = Protocolo.partir(resp);
        if (!c[0].equals(Protocolo.REGISTRO)) return out;
        if (c[1].isEmpty()) return out;
        for (String d : c[1].split(Protocolo.SEP_ENTRADAS, -1)) out.add(d);
        return out;
    }

    // Envia un mensaje probando nodos hasta dar con el lider (siguiendo REDIRECT).
    // En cada intento se recalcula el orden para priorizar al lider conocido, asi
    // que un REDIRECT en el intento N ya se aprovecha en el intento N+1.
    private String pedir(String mensaje) {
        for (int intento = 0; intento < nodos.size() * 2 + 4; intento++) {
            // arma el orden con el lider conocido primero
            List<Par> orden = new ArrayList<>();
            if (ultimoLider != null) orden.add(ultimoLider);
            for (Par p : nodos) if (ultimoLider == null || p.id != ultimoLider.id) orden.add(p);

            boolean redirigido = false;
            for (Par p : orden) {
                String resp = enviar(p, mensaje);
                if (resp == null) continue;
                String[] c = Protocolo.partir(resp);
                if (c[0].equals(Protocolo.REDIRECT)) {
                    Par lider = porDireccion(c[1]);
                    if (lider != null && (ultimoLider == null || lider.id != ultimoLider.id)) {
                        ultimoLider = lider;
                        redirigido = true;
                        break; // reintento de una vez contra el lider correcto
                    }
                    // REDIRECT a "?" o al mismo: sigo probando otros nodos
                    continue;
                }
                ultimoLider = p;
                return resp;
            }
            if (!redirigido) dormir(80); // el cluster puede estar en eleccion
        }
        return null;
    }

    private Par porDireccion(String dir) {
        for (Par p : nodos) if (p.direccion().equals(dir)) return p;
        return null;
    }

    private String enviar(Par p, String mensaje) {
        try (Socket s = new Socket()) {
            s.connect(new InetSocketAddress(p.host, p.puerto), 150);
            s.setSoTimeout(400);
            BufferedWriter out = new BufferedWriter(new OutputStreamWriter(s.getOutputStream()));
            BufferedReader in = new BufferedReader(new InputStreamReader(s.getInputStream()));
            out.write(mensaje); out.write("\n"); out.flush();
            return in.readLine();
        } catch (IOException e) {
            return null;
        }
    }

    private void dormir(long ms) {
        try { Thread.sleep(ms); } catch (InterruptedException ignore) {}
    }

    public static List<Par> parsearNodos(String[] specs, int desde) {
        List<Par> nodos = new ArrayList<>();
        for (int i = desde; i < specs.length; i++) {
            String[] p = specs[i].split(":");
            nodos.add(new Par(Integer.parseInt(p[0]), p[1], Integer.parseInt(p[2])));
        }
        return nodos;
    }
}

File: mobile/app/src/main/java/com/example/vigilante/ClienteCluster.kt
package com.example.vigilante

import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.net.InetSocketAddress
import java.net.Socket
import java.nio.charset.StandardCharsets

// Cliente del cluster Raft por sockets TCP nativos (sin frameworks). Es el mismo
// protocolo que usan el Vigilante de escritorio (Java) y el Servidor de Testeo
// (Python): una linea por mensaje, campos separados por |. Aca solo leemos el
// registro (LEER_REGISTRO). Maneja el REDIRECT al lider igual que los otros.
//
// Este cliente se reusa del patron de sockets de la PC4 (ChatClient): abrir
// socket, escribir una linea, leer la respuesta, cerrar. La lectura de red NUNCA
// va en el hilo principal de Android (ver VigilanteViewModel / MainActivity).

// una deteccion parseada del registro: tipo,imagenRef,fechaHora,camara
data class Deteccion(
    val tipo: String,
    val imagenRef: String,
    val fechaHora: String,
    val camara: String
)

class ClienteCluster(private val nodos: List<Triple<Int, String, Int>>) {

    // recordamos al ultimo lider para no reintentar desde cero
    private var ultimoLider: Triple<Int, String, Int>? = null

    // Lee el registro cometido del cluster. Devuelve la lista de detecciones,
    // o null si no se pudo contactar a nadie (cluster caido o en eleccion).
    fun leerRegistro(): List<Deteccion>? {
        val resp = pedir("LEER_REGISTRO") ?: return null
        val c = resp.split("|")
        if (c.isEmpty() || c[0] != "REGISTRO") return emptyList()
        if (c.size < 2 || c[1].isEmpty()) return emptyList()
        return c[1].split(";").mapNotNull { parsearDeteccion(it) }
    }

    private fun parsearDeteccion(s: String): Deteccion? {
        val campos = s.split(",")
        if (campos.isEmpty()) return null
        return Deteccion(
            tipo = campos.getOrElse(0) { "?" },
            imagenRef = campos.getOrElse(1) { "" },
            fechaHora = campos.getOrElse(2) { "" },
            camara = campos.getOrElse(3) { "" }
        )
    }

    // Envia un mensaje probando nodos hasta dar con el lider (siguiendo REDIRECT).
    // Recalcula el orden en cada intento para priorizar al lider conocido.
    private fun pedir(mensaje: String): String? {
        val intentosMax = nodos.size * 2 + 4
        for (intento in 0 until intentosMax) {
            val orden = ArrayList<Triple<Int, String, Int>>()
            ultimoLider?.let { orden.add(it) }
            for (n in nodos) {
                if (ultimoLider == null || n.first != ultimoLider!!.first) orden.add(n)
            }

            var redirigido = false
            for (n in orden) {
                val resp = enviar(n.second, n.third, mensaje) ?: continue
                val c = resp.split("|")
                if (c[0] == "REDIRECT") {
                    val lider = porDireccion(c.getOrElse(1) { "" })
                    if (lider != null &&
                        (ultimoLider == null || lider.first != ultimoLider!!.first)) {
                        ultimoLider = lider
                        redirigido = true
                        break
                    }
                    continue
                }
                ultimoLider = n
                return resp
            }
            if (!redirigido) {
                try { Thread.sleep(80) } catch (e: InterruptedException) { return null }
            }
        }
        return null
    }

    private fun porDireccion(dir: String): Triple<Int, String, Int>? {
        for (n in nodos) if ("${n.second}:${n.third}" == dir) return n
        return null
    }

    // Abre socket, manda una linea, lee la respuesta, cierra. Timeouts cortos
    // para que un nodo caido no cuelgue la app.
    private fun enviar(host: String, puerto: Int, mensaje: String): String? {
        var socket: Socket? = null
        return try {
            socket = Socket()
            socket.connect(InetSocketAddress(host, puerto), 300)
            socket.soTimeout = 600
            val out = BufferedWriter(
                OutputStreamWriter(socket.getOutputStream(), StandardCharsets.UTF_8))
            val input = BufferedReader(
                InputStreamReader(socket.getInputStream(), StandardCharsets.UTF_8))
            out.write(mensaje)
            out.write("\n")
            out.flush()
            input.readLine()
        } catch (e: Exception) {
            null
        } finally {
            try { socket?.close() } catch (e: Exception) {}
        }
    }

    companion object {
        // parsea specs "id:host:puerto" separados por coma o espacio
        fun parsearNodos(texto: String): List<Triple<Int, String, Int>> {
            return texto.split(Regex("[,\\s]+"))
                .filter { it.isNotBlank() }
                .mapNotNull { spec ->
                    val p = spec.split(":")
                    if (p.size != 3) return@mapNotNull null
                    val id = p[0].toIntOrNull() ?: return@mapNotNull null
                    val puerto = p[2].toIntOrNull() ?: return@mapNotNull null
                    Triple(id, p[1], puerto)
                }
        }
    }
}
